overlord/snapstate/component.go

```diff
@@ -30,8 +31,189 @@ import (
 	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/store"
+	"github.com/snapcore/snapd/strutil"
 )
 
+// InstallComponents installs all of the components in the given names list. The
+// snap represented by info must already be installed, and all of the components
+// in names should not be installed prior to calling this function.
+//
+// TODO:COMPS: respect the transaction that is passed to this function
+func InstallComponents(ctx context.Context, st *state.State, names []string, info *snap.Info, opts Options) ([]*state.TaskSet, error) {
+	var snapst SnapState
+	err := Get(st, info.InstanceName(), &snapst)
+	if err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return nil, &snap.NotInstalledError{Snap: info.InstanceName()}
+		}
+		return nil, err
+	}
+
+	currentComps, err := snapst.CurrentComponentInfos()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, comp := range currentComps {
+		if strutil.ListContains(names, comp.Component.ComponentName) {
+			return nil, fmt.Errorf("cannot install already installed component: %q", comp.Component)
```

Comment:
do we need a dedicated error for this? we have a dedicated error for snaps for the same kind of issue?